File: companies/BitGo/159.longest-substring-with-at-most-two-distinct-characters.go
package leetcode

// LengthOfLongestSubstringTwoDistinct159 is a sliding window with a character
// count map. The left side shrinks only while the window has more than two
// distinct characters.
//
// Time: O(n). Space: O(1) for bounded alphabet, O(k) generally.
func LengthOfLongestSubstringTwoDistinct159(s string) int {
	counts := map[byte]int{}
	left, best := 0, 0
	for right := 0; right < len(s); right++ {
		counts[s[right]]++
		for len(counts) > 2 {
			old := s[left]
			counts[old]--
			if counts[old] == 0 {
				delete(counts, old)
			}
			left++
		}
		best = maxInt(best, right-left+1)
	}
	return best
}

File: companies/Bloomberg/159.longest-substring-with-at-most-two-distinct-characters.java
import java.util.*;

/**
 * Algorithm:
 * Sliding window with a character count map. Expand right, and while more than
 * two distinct characters are present, shrink from the left.
 *
 * Java data structures:
 * HashMap<Character, Integer> stores counts in the current window.
 *
 * Complexity:
 * Time O(n), space O(1) because at most three character keys exist before
 * shrinking.
 */
class Solution {
    public int lengthOfLongestSubstringTwoDistinct(String s) {
        Map<Character, Integer> counts = new HashMap<>();
        int left = 0;
        int best = 0;
        for (int right = 0; right < s.length(); right++) {
            char ch = s.charAt(right);
            counts.put(ch, counts.getOrDefault(ch, 0) + 1);
            while (counts.size() > 2) {
                char old = s.charAt(left++);
                int next = counts.get(old) - 1;
                if (next == 0) {
                    counts.remove(old);
                } else {
                    counts.put(old, next);
                }
            }
            best = Math.max(best, right - left + 1);
        }
        return best;
    }
}

